### PR TITLE
edit user roles

### LINKED ISSUE #1388: UserList: Edit user
- [x] profile picture
- [x] info
- [x] roles

### PR DESCRIPTION
**Pull request by rfontanarosa**
closes #1388

https://user-images.githubusercontent.com/10047945/182178763-baa46f66-3812-41d3-b8d4-85bf5ddbaaa2.mov

### CODE REVIEW

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  unselectableCountries: Array<string>
+  role: RoleName | null
+}
+
+type Props = {
+  onChange: (value: Array<Partial<UserRole<RoleName>>>, key: string) => void
+  user: User
+}
+
+const CountryRole = (props: Props) => {
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:33]: const CountryRoles:React.FC<Props> = (props) => { 
....

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import './CountryRoles.scss'
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:1]: add styles to CountryRoles.scss

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  const secondaryRegions = useSecondaryRegion()
+  const initialModalState: ModalOptionsProps = {
+    open: false,
+    initialSelection: [],
+    unselectableCountries: [],
+    role: null,
+  }
+  const [modalOptions, setModalOptions] = useState<ModalOptionsProps>(initialModalState)
+
+  const _onClose = (selection: Array<string>, role: RoleName) => {
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:49]: useCallback

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        role,
+        assessmentId: assessment.id,
+        cycleUuid: cycle.uuid,
+      })
+    )
+
+    onChange([...user.roles.filter(({ role: _role }: UserRole<RoleName>) => _role !== role), ...selectedRoles], 'roles')
+  }
+
+  const _toggleAdmin = () => {
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:64]: useCallback

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            setModalOptions({
+              open: true,
+              initialSelection,
+              unselectableCountries,
+              role,
+            })
+          }
+
+          return (
+            <div key={role} className="edit-user__form-field-role-container validation-error-sensitive-field">
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:109]: extract it to CountryRole

Comment (rfontanarosa) [reply client/components/EditUserForm/CountryRoles/CountryRoles.tsx:117]: what do you mean?

Comment (minotogna) [reply client/components/EditUserForm/CountryRoles/CountryRoles.tsx:117]: to create an inner Component

File: server/repository/public/user/update.ts [hunk 6]
```diff
@@ -28,6 +28,19 @@ export const update = async (
...
+  const userRolePromises = roles
+    .filter((userRole) => !!userRole)
+    .map((userRole: UserRole<RoleName>) =>
+      client.query(
+        `insert into users_role (user_id, assessment_id, cycle_uuid, country_iso, role, accepted_at) values ($1, $2, $3, $4, $5, now())`,
+        [id, userRole.assessmentId, userRole.cycleUuid, userRole.countryIso, userRole.role]
+      )
+    )
+
+  await Promise.all(userRolePromises)
```
Comment (minotogna) [inline server/repository/public/user/update.ts:42]: we need to ensure only admin can updates the roles. add this check in the controller.

Plus, we need to handle the 'deactivate' functionality.

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/components/EditUserForm/CountryRoles/CountryRoles.tsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  role: RoleName | null
+}
+
+type CountryRoleProps = {
+  onClick: MouseEventHandler<HTMLButtonElement>
+  role: RoleName
+  user: User
+}
+
+const CountryRole: React.FC<CountryRoleProps> = ({ onClick, role, user }) => {
```
Comment (minotogna) [inline client/components/EditUserForm/CountryRoles/CountryRoles.tsx:34]: preferibile spostarlo nel suo proprio file

File: server/repository/public/user/update.ts [hunk 8]
```diff
@@ -28,6 +29,21 @@ export const update = async (
     [institution, lang, name, status, position, email, id]
   )
 
+  if (isAdministrator) {
```
Comment (minotogna) [inline server/repository/public/user/update.ts:32]: per renderlo più error free, creiamo un metodo esterno updateRoles che viene invocato dal Controller se è administrator

**Review by minotogna — CHANGES_REQUESTED** [review]
c'è un problema di compilazione in server/controller/user/acceptInvitaion

File: server/controller/user/update.ts [hunk 9]
```diff
@@ -2,18 +2,25 @@ import { User } from '@meta/user'
...
     profilePicture?: Express.Multer.File | null
+    isAdministrator?: boolean
   },
   client: BaseProtocol = DB
 ): Promise<User> => {
-  const { user, profilePicture } = props
+  const { user, profilePicture, isAdministrator = false } = props
 
   return client.tx(async (t) => {
     // TODO: Add activity log entry (public schema?)
```
[the remark this answers is not in the document]
Comment (rfontanarosa) [reply server/controller/user/update.ts:18]: at the moment the activity works only with the assessment schema, but the users and the roles are stored inside public schema

**Review by minotogna — APPROVED** [review]

**Pull request merged**